### PR TITLE
Feature/agency reconfiguration

### PR DESCRIPTION
**Pull request by kvahed**
Added _api/agency/add-server
Added _api/agency/remove-server

Agent's new functionality in ::reconfigure
AgentConfiguration with new functionality
State persists new configuration and immediately modifies the local.

### CODE REVIEW

**Comment by jsteemann** [comment]
Update: confirmed to be still a useful branch, and still work in progress.

**Comment by ewoutp** [comment]
Probably this is the feature needed to solve this issue: https://github.com/arangodb-helper/arangodb/issues/76

File: arangod/Agency/AgencyCommon.h [hunk 1]
```diff
@@ -40,6 +40,8 @@ static std::string const NO_LEADER = "";
 
 enum role_t { FOLLOWER, CANDIDATE, LEADER };
 
+enum apply_ret_t {APPLIED, PRECONDITION_FAILED, FORBIDDEN, UNKNOWN_ERROR};
```
Comment (neunhoef) [inline arangod/Agency/AgencyCommon.h]: Was erlaube?

Comment (neunhoef) [inline arangod/Agency/AgencyCommon.h]: Blabla

Comment (kvahed) [reply arangod/Agency/AgencyCommon.h]: This is a means to allow application of requests to the state machine to respond on a more fine grained level that just true/false

File: arangod/Agency/Agent.cpp [hunk 2]
```diff
@@ -806,12 +810,28 @@ void Agent::load() {
...
+
   MUTEX_LOCKER(tiLocker, _tiLock);
-  size_t good = 0;
   
-  std::string const myid = id();
+  size_t good = 0;
+  std::string const myid = id();  
   
   for (auto const& i : _lastAcked) {
     if (i.first != myid) {  // do not count ourselves
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp]: This still counts stuff in _lastAcked from servers which might no longer be active. Or are these entries deleted elsewhere. Note however, that this must all happen only under the _tiLock. No other locks may be acquired!

File: arangod/Agency/Agent.cpp [hunk 3]
```diff
@@ -1686,8 +1705,101 @@ void Agent::setPersistedState(VPackSlice const& compaction) {
...
 
+bool Agent::haveJoinConfig() const {
+  MUTEX_LOCKER(guard, _joinLock);
+  return (_joinConfig != nullptr);
+}
+
+void Agent::join(query_t const& config) {
+
+  if (config != nullptr) {
+    _joinConfig = config;
```
Comment (mpoeter) [inline arangod/Agency/Agent.cpp]: `_joinConfig` is not atomic, so this is a potential data race since it is outside the lock.

**Comment by kvahed** [comment]
Do not delete the original branch, please.

**Pull request closed without merging**